### DISCUSSION TITLE
Layered Architecture의 장점

### DISCUSSION THREAD

**Discussion by GIVEN53** [2024-woowacourse-study/level-interview#121] [Spring 뒷고기]
.

**Comment by jimi567** [comment]
레이어드 아키텍처(Layered Architecture)는 소프트웨어 시스템을 계층으로 나누어 각 계층이 특정한 역할과 책임을 가지도록 설계하는 아키텍처 패턴입니다

계층은 크게 프레젠테이션 레이어, 애플리케이션 레이어, 도메인 레이어, 영속 레이어로 나누어지는게 일반적입니다. 

레이어드 아키텍처로 프로젝트를 구성하게 될 경우 다음과 같은 장점을 얻을 수 있습니다.

1. 모듈화 : 각 계층은 독립적으로 개발 및 테스트가 가능해져 시스템의 복잡성을 낮출 수 있습니다
2. 유지보수 성 : 각 계층이 다른 계층에 영향을 최대한 미치지 않도록 설계 되어 있으므로( 계층별 결합도가 낮으므로) 유지보수가 용이합니다.
3. 재사용성 : 각 계층이 재사용 가능한 컴포넌트로 구성될 수 있습니다. 다른 프로젝트나 시스템에서 재사용 할 수 있는 가능성이 높아집니다.
4. 테스트 용이성 : 계층별로 단위 테스트 및 슬라이스 테스트를 작성할 수 있습니다.

**Reply by alstn113** [reply]
계층 설명이 잘못된 것 같아요.

Presentation Layer(표현 계층)
Business Layer(비즈니스 계층)
Persistence Layer(영속성 계층)
Database Layer(데이터베이스 계층)

참고
- https://prolog.techcourse.co.kr/essay-answers/101
- https://ksh-coding.tistory.com/92

**Reply by 3Juhwan** [reply]
질문이 있는데, 레어이드 아키텍처 구성이 하나로 정해져 있나요? 
저는 미션에서 `프레젠테이션 계층` - `애플리케이션 계층` - `도메인 계층` - `인프라 계층`으로 사용했어요. 
정해진 정답은 없다고 생각하지만, 구름이 언급한 부분은 확실히 일리가 있군요!

**Reply by alstn113** [reply]
그러게요. 저도 모르겠네요. 
4-tier 아키텍처를 검색하니까 대부분 저게 나오길래 하나인줄 알았어요.

방금 햄버거 아키텍처를 만들었어요. 빵 - 야채 - 패티 - 빵 
이런 식으로 많을 것 같아요. 죄송합니다,,,

**Comment by GIVEN53** [comment]
Layered Architecture는 소프트웨어 시스템을 논리적인 계층으로 나누어 구성하는 아키텍처 스타일입니다.
각 계층은 특정한 책임을 가지며, 계층 간의 의존성을 통해 시스템을 구성합니다. 이러한 구조는 여러 가지 장점을 제공합니다.

#### 1. 모듈화
시스템이 여러 계층으로 나뉘어 있어, 각각의 계층을 독립적으로 개발, 테스트할 수 있습니다.

#### 2. 예측할 수 있는 변경 전파
특정 계층의 변경은 상위 계층으로만 전파됩니다. 예를 들어, Business 계층의 변경은 Presentation 계층에만 영향을 줍니다. 따라서 상위 계층은 하위 계층을 믿고 활용할 수 있습니다.

#### 3. 명확한 역할 분리
각 계층이 특정 역할과 책임을 가지므로, 시스템의 각 부분이 어떤 기능을 하는지 명확히 이해할 수 있습니다.

#### 4. 테스트 용이성
계층별로 독립적인 테스트가 가능하기 때문에, 테스트 범위를 좁혀 효과적으로 테스트할 수 있습니다.

#### 5. 재사용성 향상
특정 계층의 컴포넌트는 다른 프로젝트에서도 재사용할 수 있습니다. 예를 들어, 데이터 액세스 계층은 여러 프로젝트에서 동일한 데이터베이스 접근 로직을 공유할 수 있습니다.

#### 6. 유지보수
특정 계층의 기술 스택이나 구현 방법을 변경하더라도 다른 계층에 미치는 영향이 적어 유지보수가 용이합니다.
